test/DefaultCluster.Tests/GrainReferenceTest.cs

```diff
@@ -102,6 +103,33 @@ public void GrainReference_Json_Serialization()
         }
 
         [Fact, TestCategory("Serialization"), TestCategory("JSON")]
+        public async Task GrainReference_Json_Serialization_Nested()
+        {
+            var settings = OrleansJsonSerializer.GetDefaultSerializerSettings(HostedCluster.SerializationManager, HostedCluster.GrainFactory);
+            
+            var grain = HostedCluster.GrainFactory.GetGrain<ISimpleGrain>(GetRandomGrainId());
+            await grain.SetA(56820);
+            var input = new GenericGrainReferenceHolder
+            {
+                Reference = grain
+            };
+
+            var json = JsonConvert.SerializeObject(input, settings);
+            var output = JsonConvert.DeserializeObject<GenericGrainReferenceHolder>(json, settings);
+
+            Assert.Equal(input.Reference, output.Reference);
+            var reference = output.Reference.Cast<ISimpleGrain>();
```

Comment:
In the updated version, I use a C# cast to prove that it's an instance of `ISimpleGrain` at runtime, rather than a generic `GrainReference` which is then converted to an `ISimpleGrain` ref.